DTensor: C++ compute_global_tensor_info

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #163667
* __->__ #162990

compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

swolchok committed:
DTensor: C++ compute_global_tensor_info

compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
+            placement.attr(dtensor_interned_strings.is_shard)().ptr())) {
+      const auto shard_dim =
+          py::cast<int64_t>(placement.attr(dtensor_interned_strings.dim));
+      TORCH_CHECK(
```
Don't we have a TORCH_CHECK_ASSERT to raise an Assertion Error?

swolchok replied:
not that I found. TORCH_INTERNAL_ASSERT exists but doesn't seem to do anything differently w.r.t. the exception it raises.

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
+      .ptr();
+}
+
+static const char compute_global_tensor_info_doc[] =
```
Make it constexpr too

Skylion007 commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1122,6 +1135,124 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
+    PyObject* const* args,
+    Py_ssize_t nargs) {
+  HANDLE_TH_ERRORS
+  TORCH_CHECK(
```
TORCH_CHECK_VALUE

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Python 3.13 renamed _PyCFunctionFast to remove the underscore on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
parenthesize python_compat macros on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok commented:
> will try TORCH_LIBRARY

Hmm, making this a custom op seems bad, at least right now. The function signature includes DeviceMesh (not currently IValue-able) and a list of Placements (also not currently IValue-able). I guess maybe I'm supposed to 1) add the Placement stuff from https://github.com/pytorch/pytorch/pull/163030 to IValue 2) have a C++ DeviceMesh representation 3) add that DeviceMesh representation to IValue 4) teach the custom operator / native function schema stuff about all that and then 5) sequence this PR after all those things?

On the other hand, torch.compiler.allow_in_graph might make this easy. I'll try that first.

swolchok commented:
> On the other hand, torch.compiler.allow_in_graph might make this easy. I'll try that first.

Simply adding
```
        import torch._C
        torch.compiler.allow_in_graph(torch._C._DTensor_compute_global_tensor_info)
```

to the start of test_dtensor_dynamic does not cause it to pass. @ezyang does it sound right that C++ifying utility functions that happen to work with Tensors has to be done through custom operators, which necessitates making all their argument types (i.e., DeviceMesh, Placement) go through IValue?

swolchok committed:
reorder #162990 after #163030 given current custom op uncertainty on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix lintrunner mypy on 162990 on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix missing torch.distributed.fsdp import in test that used torch.ops.fsdp on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
rebase on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

ezyang commented:
They look ignorable. Is there a reason you put periodic label on it?

ezyang commented on test/inductor/test_distributed_patterns.py:
```diff
@@ -3,6 +3,7 @@
 import functools
 
 import torch
+import torch.distributed.fsdp
```
weird

ezyang commented on torch/csrc/utils/python_compat.h:
```diff
@@ -9,11 +9,11 @@ extern "C" {
+#define IS_PYTHON_3_12_PLUS (PY_VERSION_HEX >= 0x030C0000)
+#define IS_PYTHON_3_13_PLUS (PY_VERSION_HEX >= 0x030D0000)
+#define IS_PYTHON_3_14_PLUS (PY_VERSION_HEX >= 0x030E0000)
+#define IS_PYTHON_3_15_PLUS (PY_VERSION_HEX >= 0x030F0000)
```
👍

ezyang commented on torch/csrc/utils/pycfunction_helpers.h:
```diff
@@ -11,3 +12,15 @@ inline PyCFunction castPyCFunctionWithKeywords(PyCFunctionWithKeywords func) {
 }
+
+#if !IS_PYTHON_3_13_PLUS
+using PyCFunctionFast = _PyCFunctionFast;
```
cc @williamwen42

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
+  TORCH_CHECK_TYPE(
+      THPVariable_Check(args[0]),
+      "compute_global_tensor_info 1st argument must be Tensor!");
+  const auto& tensor = THPVariable_Unpack(args[0]);
```
ah the joy of const reference lifetime extension

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
+    // calling back into Python.
+    const auto& cpp_placement = placement.cast<const distributed::Placement&>();
+    if (const auto* cpp_shard =
+            dynamic_cast<const distributed::Shard*>(&cpp_placement)) {
```
I love the smell of RTTI for breakfast (do we really not have a type tag on these yet?)

swolchok replied:
Hmm. It's been long enough that I'm not 100% sure what I was thinking, but.I now note that if this was just an  `is_shard()` call and someone has written a Python class that is a subclass of Placement but not the C++ shard class and yet is_shard returns true for it, sadness would ensue

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
+  // apparently we can't rely on the bound method to stick around.
+  py::object mesh_size;
+  for (const auto& placement : placements) {
+    // TODO: C++ify DeviceMesh somehow; profiling seems
```
cc @fduwjj @lw 

In this case, you don't actually need the full mesh: you only need the size which is basically an IntArrayRef.

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1130,6 +1140,130 @@ static PyObject* DTensor_OpSchema_post_init(PyObject* mod, PyObject* self) {
+      TORCH_CHECK(
+          shard_dim >= 0,
+          "Shard placements should have negative dims normalized in the user-facing APIs: ",
+          py::cast<std::string>(py::str(placement)));
```
this is annoyingly wordy, printing placement directly ought to work

swolchok committed:
rebase to pick up reapply of dependency on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix bottom-of-stack PR on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix lint and missing import, respond to a couple review comments on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
Update on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix tests, hopefully on "DTensor: C++ compute_global_tensor_info"


compute_global_tensor_info is on the hot path for DTensor.{from,to}_local. More incremental progress toward C++.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

swolchok commented:
failures this time around:
-  [linux-jammy-cuda12.8-py3.10-gcc11 / test (default, 4, 5, linux.g6.4xlarge.experimental.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/18918352725/job/54010750231?pr=162990#logs) looks like it was also segfaulting on main around the time of the base commit: https://github.com/pytorch/pytorch/actions/runs/18862532009/job/53825651259
- the executorch job is marked unstable and seems quite unlikely to be related


as best I can tell we're good to merge with -i.

swolchok commented:
@pytorchbot merge -i

swolchok committed:
Update

[ghstack-poisoned]

jithunnair-amd commented:
@swolchok You might want to use `merge -f` if the ROCm distributed jobs are the only ones pending, since the ROCm distributed jobs running in the latest round are looking for the label `linux.rocm.gpu.mi250.4`, and will probably be queued for a long time.
